src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php

```diff
@@ -0,0 +1,268 @@
+        ) {
+            return $propertyMetadata->withSchema($propertySchema);
+        }
+
+        $valueSchema = [];
+        foreach ($types as $type) {
+            if ($isCollection = $type->isCollection()) {
+                $keyType = $type->getCollectionKeyTypes()[0] ?? null;
+                $valueType = $type->getCollectionValueTypes()[0] ?? null;
+            } else {
+                $keyType = null;
+                $valueType = $type;
+            }
+
+            if (null === $valueType) {
+                $builtinType = 'string';
+                $className = null;
+            } else {
+                $builtinType = $valueType->getBuiltinType();
+                $className = $valueType->getClassName();
+            }
+
+            $propertyType = $this->getType(new Type($builtinType, $type->isNullable(), $className, $isCollection, $keyType, $valueType), $propertyMetadata->isReadableLink());
+            if (!\in_array($propertyType, $valueSchema, true)) {
+                $valueSchema[] = $propertyType;
+            }
+        }
+
+        // only one builtInType detected (should be "type" or "$ref")
+        if (1 === \count($valueSchema)) {
+            return $propertyMetadata->withSchema($propertySchema + $valueSchema[0]);
+        }
+
+        // multiple builtInTypes detected: determine oneOf/allOf if union vs intersect types
+        try {
+            $reflectionClass = new \ReflectionClass($resourceClass);
+            $reflectionProperty = $reflectionClass->getProperty($property);
+            $composition = $reflectionProperty->getType() instanceof \ReflectionUnionType ? 'oneOf' : 'allOf';
+        } catch (\ReflectionException) {
+            // cannot detect types
+            $composition = 'anyOf';
+        }
+
+        return $propertyMetadata->withSchema($propertySchema + [$composition => $valueSchema]);
+    }
+
+    private function getType(Type $type, ?bool $readableLink = null): array
+    {
+        if ($type->isCollection()) {
+            $keyType = $type->getCollectionKeyTypes()[0] ?? null;
+            $subType = ($type->getCollectionValueTypes()[0] ?? null) ?? new Type($type->getBuiltinType(), false, $type->getClassName(), false);
+
+            if (null !== $keyType && Type::BUILTIN_TYPE_STRING === $keyType->getBuiltinType()) {
+                return $this->addNullabilityToTypeDefinition([
+                    'type' => 'object',
+                    'additionalProperties' => $this->getType($subType, $readableLink),
+                ], $type);
+            }
+
+            return $this->addNullabilityToTypeDefinition([
+                'type' => 'array',
+                'items' => $this->getType($subType, $readableLink),
+            ], $type);
+        }
+
+        return $this->addNullabilityToTypeDefinition($this->makeBasicType($type, $readableLink), $type);
```

Comment:
if (!$type->isCollection()) {
      return $this->addNullabilityToTypeDefinition($this->makeBasicType($type, $readableLink), $type);
    }

    $keyType = $type->getCollectionKeyTypes()[0] ?? null;
    $subType = ($type->getCollectionValueTypes()[0] ?? null) ?? new Type($type->getBuiltinType(), false, $type->getClassName(), false);

    if (null !== $keyType && Type::BUILTIN_TYPE_STRING === $keyType->getBuiltinType()) {
      return $this->addNullabilityToTypeDefinition([
          'type' => 'object',
          'additionalProperties' => $this->getType($subType, $readableLink),
      ], $type);
    }

    return $this->addNullabilityToTypeDefinition([
        'type' => 'array',
        'items' => $this->getType($subType, $readableLink),
    ], $type);
```